atintegrators/ExactHamiltonianPass.cc

```diff
@@ -0,0 +1,211 @@
+                               Elem->T1, Elem->T2, Elem->R1, Elem->R2,
+                               Elem->MaxOrder, Elem->NumIntSteps,
+                               Elem->Phi, Elem->Type, Elem->gK,
+                               Elem->MultipoleFringe, num_particles);
+
+    return Elem;
+}
+
+MODULE_DEF(ExactHamiltonianPass) /* Dummy module initialisation */
+
+#endif /* defined(PYAT) || defined(MATLAB_MEX_FILE) */
+
+
+#if defined(MATLAB_MEX_FILE)
+void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
+{
+    if (nrhs >= 2) {
+        double * r_in;
+        const mxArray * ElemData = prhs[0];
+        int num_particles = mxGetN(prhs[1]);
+        int m = mxGetM(prhs[1]);
+        if (m!=6)
+            mexErrMsgTxt("Second argument must be a 6 x N matrix");
+
+        double le, bending_angle;
+        double *polynom_a, *polynom_b;
+        long max_order, num_int_steps, type, multipole_fringe;
+        double *R1, *R2, *T1, *T2;
+        double phi, gK;
+        le = atGetDouble(ElemData,"Length"); check_error();
+        bending_angle = atGetDouble(ElemData,"BendingAngle"); check_error();
+        polynom_a = atGetDoubleArray(ElemData,"PolynomA"); check_error();
+        polynom_b = atGetDoubleArray(ElemData,"PolynomB"); check_error();
+        max_order = atGetLong(ElemData, "MaxOrder"); check_error();
+        num_int_steps = atGetLong(ElemData, "NumIntSteps"); check_error();
+        type = atGetLong(ElemData, "Type"); check_error();
+        /*optional fields*/
+        multipole_fringe = atGetOptionalLong(ElemData, "MultipoleFringe", 0); check_error();
+        phi = atGetOptionalDouble(ElemData,"Phi", 0.0); check_error();
+        gK = atGetOptionalDouble(ElemData,"gK", 0.0); check_error();
+        R1=atGetOptionalDoubleArray(ElemData,"R1"); check_error();
+        R2=atGetOptionalDoubleArray(ElemData,"R2"); check_error();
+        T1=atGetOptionalDoubleArray(ElemData,"T1"); check_error();
+        T2=atGetOptionalDoubleArray(ElemData,"T2"); check_error();
+
+        /* ALLOCATE memory for the output array of the same size as the input  */
+        plhs[0] = mxDuplicateArray(prhs[1]);
+        r_in = mxGetPr(plhs[0]);
+        ExactHamiltonianPass(r_in, le, polynom_a, polynom_b, T1, T2, R1, R2, max_order,
+                     num_int_steps, phi, type, gK, multipole_fringe, num_particles);
+    }
+    else if (nrhs == 0) {
+        /* list of required fields */
+        plhs[0] = mxCreateCellMatrix(5,1);
+        mxSetCell(plhs[0],0,mxCreateString("Length"));
+        mxSetCell(plhs[0],0,mxCreateString("BendingAngle"));
+        mxSetCell(plhs[0],1,mxCreateString("PolynomA"));
+        mxSetCell(plhs[0],2,mxCreateString("PolynomB"));
+        mxSetCell(plhs[0],3,mxCreateString("MaxOrder"));
+        mxSetCell(plhs[0],4,mxCreateString("NumIntSteps"));
+        mxSetCell(plhs[0],4,mxCreateString("Type"));
+        mxSetCell(plhs[0],4,mxCreateString("MultipoleFringe"));
```

Comment:
Since MultipoleFringe, Phi and gK are now back to optional, they should be moved to the 2nd section.